pkg/detectors/salesforce/salesforce.go

```diff
@@ -0,0 +1,118 @@
+package salesforce
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/trufflesecurity/trufflehog/v3/pkg/common"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/detectors"
+	"github.com/trufflesecurity/trufflehog/v3/pkg/pb/detectorspb"
+)
+
+const (
+	currentVersion = "58.0" // current Salesforce version
+)
+
+type Scanner struct {
+	client *http.Client
+}
+
+// Ensure the Scanner satisfies the interface at compile time.
+var _ detectors.Detector = (*Scanner)(nil)
+
+var (
+	defaultClient = common.SaneHttpClient()
+	// Make sure that your group is surrounded in boundary characters such as below to reduce false positives.
+	accessTokenPat = regexp.MustCompile(`\b00[a-zA-Z0-9]{13}![a-zA-Z0-9_.]{96}\b`)
+	instancePat    = regexp.MustCompile(`\bhttps://[0-9a-zA-Z-\.]{1,100}\.my\.salesforce\.com\b`)
+)
+
+// Keywords are used for efficiently pre-filtering chunks.
+// Use identifiers in the secret preferably, or the provider name.
+func (s Scanner) Keywords() []string {
+	return []string{"salesforce"}
+}
+
+// FromData will find and optionally verify Salesforce secrets in a given set of bytes.
+func (s Scanner) FromData(ctx context.Context, verify bool, data []byte) (results []detectors.Result, err error) {
+	dataStr := string(data)
+
+	instanceMatches := instancePat.FindAllStringSubmatch(dataStr, -1)
+	tokenMatches := accessTokenPat.FindAllStringSubmatch(dataStr, -1)
+
+	for _, instance := range instanceMatches {
+		if len(instance) != 1 {
+			continue
+		}
+
+		instanceMatch := strings.TrimSpace(instance[0])
+
+		for _, token := range tokenMatches {
+			if len(token) != 1 {
+				continue
+			}
+
+			tokenMatch := strings.TrimSpace(token[0])
+
+			s1 := detectors.Result{
+				DetectorType: detectorspb.DetectorType_Salesforce,
+				Raw:          []byte(tokenMatch),
+			}
+
+			if verify {
+				client := s.client
+				if client == nil {
+					client = defaultClient
+				}
+
+				req, err := http.NewRequestWithContext(ctx, "GET", instanceMatch+"/services/data/v"+currentVersion+"/query?q=SELECT+name+from+Account", nil)
+				if err != nil {
+					continue
+				}
+				req.Header.Set("Authorization", "Bearer "+tokenMatch)
+				if err != nil {
+					continue
+				}
+
+				res, err := client.Do(req)
+
+				if err != nil {
+					// End execution, append Detector Result if request fails to prevent panic on response body checks
+					s1.VerificationError = err
+					results = append(results, s1)
+					continue
+				}
+
+				verifiedBodyResponse, err := common.ResponseContainsSubstring(res.Body, "records")
+				if err != nil {
+					return nil, err
```

Comment:
is this written like this because a failure to read the body indicates an underlying problem with the detector, rather than a problem verifying this particular candidate secret?

Reply:
yes - we could have an error reading the body. 

```
func ResponseContainsSubstring(reader io.ReadCloser, target string) (bool, error) {
	scanner := bufio.NewScanner(reader)
	for scanner.Scan() {
		if strings.Contains(scanner.Text(), target) {
			return true, nil
		}
	}
	if err := scanner.Err(); err != nil {
		return false, err
	}
	return false, nil
}
```

Reply:
but if we do, is bailing on the entire detection the correct action?

Reply:
hm what's the alternative? 

there's no network calls etc so a retry doesn't make sense afaik... figure it's best to pass on the error and call it quits instead of guessing the secret candidate is not verified & moving on.

Reply:
you could mark it as an indeterminate failure to verify

Reply:
hm i thought indeterminate was only for failed network calls? 

if `common.ResponseContainsSubstring(res.Body, "records")` returns an error, it would be a Scanner.Err() that looks like it's basically a parsing error

```
// Err returns the first non-EOF error that was encountered by the Scanner.
func (s *Scanner) Err() error {
	if s.err == io.EOF {
		return nil
	}
	return s.err
}
```

Reply:
it's just for any failure that isn't known to be caused by the candidate secret being invalid. i don't know what could cause `Scanner.Err()`, but i'm guessing it's not invalid credentials. if you record it as an indeterminate verification, we'll at least have a record of it; but in doing so we might accumulate records that we don't want for some reason. up to you!